### PR TITLE
Refactoring and cleaning up of the JS client code

### PR DESCRIPTION
**Pull request by r3trofitted**
This PR tries to clean up some of the mess in the current JS code, and lay down some groundwork for further improvements.

The main change is rather opinionated: I'd like to systematise the  use of _factory functions_ for object creation, and rely as little as possible on `this`. This is why unique objects (such as `Dockyard`) have been replaced with such factory functions (still capitalised, though, eg. `Dockyard()`).

I'd also like to systematise the _extension_ of objects from the DOM with extra, game-related properties, when needed. For example, the mothership is mostly a `<div>`, but it also needs some game mechanics. Instead of having two separate objects, one for the DOM element and one for the JS engine, and linking them (eg. through something like `mothership.element = Document.getElementById("#mothership")`), I find it best to primarily use the DOM object, and add the necessary logic to it through `Object.assign()` and `Object.defineProperties()`.

Finally, I've added several unit tests. They should be taken with a grain of salt, though, since they have been written _after_ the actual implementation, but at least we're not leaving a proverbial broken window in our code base.

There is still work to be done, mostly on the global `Battletype` object, and some logic that had to be de-abstracted could be DRYed again, but it can be deferred to a latter PR.

### CODE REVIEW

**Comment by cveneziani** [comment]
I've read everything. Let's go for it!
How can we do with the other pending PRs that are touching the JS?

File: app/assets/javascripts/battletype/combat_zone.js [hunk 1]
```diff
@@ -1,22 +1,17 @@
...
-          return Math.floor(Math.random() * this.height()) - 80;
-        }
+  var TOP_MARGIN = 15;
+  
+  this.CombatZone = function (element) {
+    var hud    = element.getElementsByClassName("hud")[0];
+    var hudTop = hud ? hud.offsetHeight : 0;
+    
+    Object.assign(element, {
+      randomFreeVerticalSlot: function () {
```
Comment (cveneziani) [inline app/assets/javascripts/battletype/combat_zone.js]: Shall we update the tests?

Comment (r3trofitted) [inline app/assets/javascripts/battletype/combat_zone.js]: It'd be better, but I dont' really know _how_, actually…

Comment (cveneziani) [inline app/assets/javascripts/battletype/combat_zone.js]: What about mocking the randomness part?

**Pull request closed without merging**